vyper/venom/analysis/mem_ssa.py

```diff
@@ -0,0 +1,388 @@
+
+    def __hash__(self) -> int:
+        return self.id
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.id_str})"
+
+
+class MemoryDef(MemoryAccess):
+    """Represents a definition of memory state"""
+
+    def __init__(self, id: int, store_inst: IRInstruction):
+        super().__init__(id)
+        self.store_inst = store_inst
+        self.loc = store_inst.get_write_memory_location()
+
+
+class MemoryUse(MemoryAccess):
+    """Represents a use of memory state"""
+
+    def __init__(self, id: int, load_inst: IRInstruction):
+        super().__init__(id)
+        self.load_inst = load_inst
+        self.loc = load_inst.get_read_memory_location()
+
+
+class MemoryPhi(MemoryAccess):
+    """Represents a phi node for memory states"""
+
+    def __init__(self, id: int, block: IRBasicBlock):
+        super().__init__(id)
+        self.block = block
+        self.operands: List[Tuple[MemoryDef, IRBasicBlock]] = []
+
+
+class MemSSA(IRAnalysis):
+    """
+    This pass converts memory/storage operations into Memory SSA form,
+    tracking memory definitions and uses explicitly.
+    """
+
+    VALID_LOCATION_TYPES = {"memory", "storage"}
+
+    def __init__(self, analyses_cache, function, location_type: str = "memory"):
+        super().__init__(analyses_cache, function)
+        if location_type not in self.VALID_LOCATION_TYPES:
+            raise ValueError(f"location_type must be one of: {self.VALID_LOCATION_TYPES}")
+        self.location_type = location_type
+        self.load_op = "mload" if location_type == "memory" else "sload"
+        self.store_op = "mstore" if location_type == "memory" else "sstore"
+
+        # Memory SSA specific state
+        self.next_id = 1  # Start from 1 since 0 will be live_on_entry
+        self.live_on_entry = MemoryAccess(0)  # live_on_entry node
+        self.memory_defs: Dict[IRBasicBlock, List[MemoryDef]] = {}
+        self.memory_uses: Dict[IRBasicBlock, List[MemoryUse]] = {}
+        self.memory_phis: Dict[IRBasicBlock, MemoryPhi] = {}
+        self.current_def: Dict[IRBasicBlock, MemoryAccess] = {}
+        self.inst_to_def: Dict[IRInstruction, MemoryDef] = {}
+        self.inst_to_use: Dict[IRInstruction, MemoryUse] = {}
+
+    def analyze(self):
+        # Request required analyses
+        self.cfg = self.analyses_cache.request_analysis(CFGAnalysis)
+        self.dom = self.analyses_cache.request_analysis(DominatorTreeAnalysis)
+        self.alias = self.analyses_cache.request_analysis(MemoryAliasAnalysis)
+
+        # Build initial memory SSA form
+        self._build_memory_ssa()
+
+        # Clean up unnecessary phi nodes
+        self._remove_redundant_phis()
+
+    def mark_location_volatile(self, loc: MemoryLocation) -> MemoryLocation:
+        volatile_loc = self.alias.mark_volatile(loc)
+
+        for bb in self.memory_defs:
+            for mem_def in self.memory_defs[bb]:
+                if self.alias.may_alias(mem_def.loc, loc):
```

Comment:
i'm not sure this is sufficient. this poisons aliasing intervals, but it does not progressively add to the worklist. for example, if `[0, 10)` is volatile and poisons `[5, 15)`, the location `[10, 20)` should also be poisoned. let me know if my understanding is correct.